File: fun/x-or-fun.c
#include <stdlib.h>
#include <stdio.h>
#include <limits.h>
#include <stdbool.h>

// Build: gcc -o xfun x-or-fun.c
// Run: execute xfun

int max(int array[], int size);
void printArray(int array[], int size);
void printBits(size_t const size, void const * const ptr);

bool showDetails = true; // Change to true to see details in output, false to hide them.

int main(int argc, char * argv[]) {
   
   int array[] = {1, 2, 3, 4, 6, 7, 8, 9, 10};
   int arraySize = sizeof(array)/sizeof(int);
	printf("\nThis tool finds a missing number from the array below,\n");
	printf("using XOR operator. The tool goes through two loops, first\n");
	printf("XORing the numbers from 1 to max value found in array,\n");
	printf("(10 in this example). Then second loop XORs the actual values\n");
	printf("in the array. The result will then contain the missing value.\n\n");

   printArray(array, arraySize);
   
   int maxValue = max(array, arraySize);
   int result = 0;
   printf("First loop...\n");
   for (int number = 1; number <= maxValue; number++) {
      if (showDetails) { printf("    "); printBits(sizeof(int), &result); }
      if (showDetails) { printf("XOR "); printBits(sizeof(int), &number); }
      result = result ^ number;
      if (showDetails) { printf("=   "); printBits(sizeof(int), &result); printf("----\n"); }
   }
   printf("Second loop...\n");
   for (int index = 0; index < arraySize; index++) {
      if (showDetails) { printf("    "); printBits(sizeof(int), &result); }
      if (showDetails) { printf("XOR "); printBits(sizeof(int), &array[index]); }
      result = result ^ array[index];
      if (showDetails) { printf("=   "); printBits(sizeof(int), &result); printf("----\n"); }
   }
   printf("Missing number is: %d\nAnd missing number in binary is: \n", result);
   printBits(sizeof(int), &result);
   return EXIT_SUCCESS;
}

int max(int array[], int size) {
   int maxValue = INT_MIN;
   for (int i = 0; i < size; i++) {
      if (array[i] > maxValue) {
         maxValue = array[i];
      }
   }
   return maxValue;
}

void printArray(int array[], int size) {
   printf("Array contents: \n");
   for (int i = 0; i < size; i++) {
      printf("%d, ", array[i]);
   }
   printf("\n");
}

void printBits(size_t const size, void const * const ptr)
{
   unsigned char *b = (unsigned char*) ptr;
   unsigned char byte;
   int i, j;
   
   for (i = size-1; i >= 0; i--) {
      for (j = 7; j >= 0; j--) {
         byte = (b[i] >> j) & 1;
         printf("%u", byte);
         if (j % 4 == 0) {
            printf(" ");
         }
      }
   }
   puts("");
}

File: fun/x-or-fun.swift
// Swift version of the x-or-fun.c.
// Build: swiftc -o xfun x-or-fun.swift
// Run: execute the xfun binary.

let array = [1,2,3,4,6,7,8,9,10]

let showDetails = true // Change to true to see what happens
var result = 0
let maxNum = array.max()!

print("\nThis tool finds a missing number from the array below,");
print("using XOR operator. The tool goes through two loops, first");
print("XORing the numbers from 1 to max value found in array,");
print("10 in this example). Then second loop XORs the actual values");
print("in the array. The result will then contain the missing value.\n");

print("Array is \(array)")
print("First loop...")
for number in 1...maxNum {
   if showDetails { print("    \(pad(string: String(result, radix: 2), toSize: 8))") }
   if showDetails { print("XOR \(pad(string: String(number, radix: 2), toSize: 8))") }
   result = result ^ number
   if showDetails { print("=   \(pad(string: String(result, radix: 2), toSize: 8))") }
   if showDetails { print("-----") }
}
print("Second loop...")
for number in array {
   if showDetails { print("    \(pad(string: String(result, radix: 2), toSize: 8))") }
   if showDetails { print("XOR \(pad(string: String(number, radix: 2), toSize: 8))") }
   result = result ^ number
   if showDetails { print("=   \(pad(string: String(result, radix: 2), toSize: 8))") }
   if showDetails { print("-----") }
}

print("Missing number is \(result), in binary: \(pad(string: String(result, radix: 2), toSize: 8))")

// Helper function for printing out details
func pad(string : String, toSize: Int) -> String {
   var padded = string
   for _ in 0..<(toSize - string.count) {
      padded = "0" + padded
   }
   return padded
}

